controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java

```diff
@@ -1956,6 +1956,12 @@ private HttpResponse submit(String tenant, String application, HttpRequest reque
 
         ApplicationPackage applicationPackage = new ApplicationPackage(dataParts.get(EnvironmentResource.APPLICATION_ZIP), true);
 
+        // if we are in public and 'security/clients.pem' is not present (i.e. we don't have any trusted
+        // certificates), we should fail early instead of waiting for the configserver to fail.
+        if (controller.system().isPublic() && applicationPackage.trustedCertificates().isEmpty()) {
```

Comment:
This belongs in `ApplicationPackageValidator`.

Reply:
Should be fixed now. Take another look. I have no idea if the test works since the build got broken by a later `git pull` :(